### DISCUSSION TITLE
Difference between printed and specified predictor matrix

### DISCUSSION THREAD

**Discussion by kasiakasia93** [amices/mice#578] [Q&A] [answered]
Hello, I am new to github discussions and mice. I have looked through the main van Buuren paper and the function documentation as well as the discussion threads and still need some help. 

My goal is to impute 2 continuous variables (bounded between 0-100) in a longitudinal data set that is in long format. The data set is unbalanced without fixed measurement occasions. 

It is my understanding that I will use the 2L.norm method to impute these variables. 

I have edited a prediction matrix to include my covariates of interest (specified at '1') and the identification variable (specified as '-2'). I have not specified any random effects. 

When I run my mice object it does output that 2l.norm has been used but the predictor matrix is different than the one if i do imp$predictorMatrix. 

Is this a bug or something wrong with my process? Which predictor matrix is being used? 

Thank you for any help! 

- Katherine

```{r}
#simpleMI
imp1 <- mice(dataset1, maxit=1,seed = 123)
```

```{r}
#edit predictor matrix for multilevel
pred<-imp1$pred
pred["Score1","id"] <- c(-2)
pred["Score2","id"] <- c(-2)
pred["Score1",c("c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(1,1,1,1,1,1,0,1,1,1,1,1,0)
pred["Score2",c("c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(1,1,1,1,1,1,0,1,1,1,1,1,0)
pred["c1",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["c2",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["c3",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["c4",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["c5",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["c6",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["c8",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["c9",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["c10",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["c11",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["c12",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)
pred["id",c("id","c1","c2","c3","c4","c5","c6","Score1","c8","c9","c10","c11","c12","Score2")]<- c(0,0,0,0,0,0,0,0,0,0,0,0,0,0)

## confirm that the matrix is what I want 
pred

```

![Screen Shot 2023-08-15 at 3 32 36 PM](https://github.com/amices/mice/assets/142345338/6927034f-6975-4b4c-bb28-13326cf4ec44)

#multi-level imputation 
```{r}

imp2 <- mice(dataset1,m=5, seed = 123, meth = c("","","","","","","","2l.norm","","","","","","2l.norm"), pred = pred, maxit=1)

imp2
# here the predictor matrix is incorrect 
```
![Screen Shot 2023-08-15 at 3 34 44 PM](https://github.com/amices/mice/assets/142345338/b6c0934e-97aa-4ada-a9fd-a44f6e1c1299)
![Screen Shot 2023-08-15 at 3 35 00 PM](https://github.com/amices/mice/assets/142345338/8f42b6fd-7afc-45d4-996e-c58b7a549899)

#but here it is correct 
imp2$predictorMatrix

![Screen Shot 2023-08-15 at 3 35 29 PM](https://github.com/amices/mice/assets/142345338/e1a92ad4-3c68-4117-8916-a95c4645162c)

mice.data<-complete(imp2)
#describe(mice.data) it also does NOT respect the bounds of the variable so I will have to use the squeeze function somehow

**Comment by stefvanbuuren** [comment, marked as answer]
Welcome to MICE discussions. 

There is a difference between the real object `imp2$predictorMatrix` and the version that is printed on the terminal by `imp2`. Here's what `mice` does to produce the printed version:

``` r
library(mice, warn.conflicts = FALSE)
mice:::print.mids
#> function (x, ...) 
#> {
#>     cat("Class: mids\n")
#>     cat("Number of multiple imputations: ", x$m, "\n")
#>     cat("Imputation methods:\n")
#>     print(x$method, ...)
#>     cat("PredictorMatrix:\n")
#>     print(head(x$predictorMatrix), ...)
#>     if (!is.null(x$loggedEvents)) {
#>         cat("Number of logged events: ", nrow(x$loggedEvents), 
#>             "\n")
#>         print(head(x$loggedEvents), ...)
#>     }
#>     invisible(x)
#> }
#> <bytecode: 0x10f1e5af8>
#> <environment: namespace:mice>
```

<sup>Created on 2023-08-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

The line  `print(head(x$predictorMatrix), ...)` prints only the first six lines in order to keep output manageable. Internally, `mice` always uses `imp2$predictorMatrix`.
